torch/csrc/autograd/python_function.cpp

```diff
@@ -480,10 +431,10 @@ static void _save_variables(THPFunction* self)
 }
 
 // Mark requires_grad = 0 on non-differentiable variables (as per non_differentiable)
-static std::unordered_set<PyObject*>
+static std::unordered_set<Variable*>
 _parse_non_differentiable(THPFunction *self)
 {
-  std::unordered_set<PyObject*> set;
+  std::unordered_set<Variable*> set;
```

Comment:
This is very unlikely to be the right thing. Variable itself is a pointer type; if you need a pointer to hash on, probably the right thing is to call `unsafeGetTensorImpl()` which will give you a `TensorImpl*` which is appropriate to hash on.

Reply:
It doesn't have to be a pointer, I can have a set of Variable instead.

Reply:
What you've done here is hash on the location of the `Variable` (i.e., `TensorImpl**`), this accidentally works because you always take an address from a PyObject which then uniquely identifies the memory location of `Variable`

Reply:
Sure, a set of Variable is good to, assuming they are hashable (I'm not sure they are. But it would be a reasonable thing for them to be hashable.)